src/mlpack/methods/ann/layer/batch_norm_impl.hpp

```diff
@@ -0,0 +1,373 @@
+        batchSize * higherDimension, false, false);
+
+    outputTemp.each_slice() -= arma::repmat(runningMean.t(),
+        inputSize, 1);
+    outputTemp.each_slice() /= arma::sqrt(arma::repmat(runningVariance.t(),
+        inputSize, 1) + eps);
+    outputTemp.each_slice() %= arma::repmat(gamma.t(),
+        inputSize, 1);
+    outputTemp.each_slice() += arma::repmat(beta.t(),
+        inputSize, 1);
+  }
+}
+
+template<typename MatType>
+void BatchNormType<MatType>::Backward(
+    const MatType& input,
+    const MatType& gy,
+    MatType& g)
+{
+  const arma::mat stdInv = 1.0 / arma::sqrt(variance + eps);
+
+  const size_t batchSize = input.n_cols;
+  const size_t inputSize = inputDimension;
+
+  arma::Cube<typename MatType::elem_type> gyTemp(
+      const_cast<MatType&>(gy).memptr(), inputSize, size,
+      batchSize * higherDimension, false, false);
+  arma::Cube<typename MatType::elem_type> gTemp(
+      const_cast<MatType&>(g).memptr(), inputSize, size,
+      batchSize * higherDimension, false, false);
+
+  // Step 1: dl / dxhat.
+  arma::Cube<typename MatType::elem_type> norm =
+      gyTemp.each_slice() % arma::repmat(gamma.t(), inputSize, 1);
+
+  // Step 2: sum dl / dxhat * (x - mu) * -0.5 * stdInv^3.
+  MatType temp = arma::sum(norm % inputMean, 2);
+  MatType vars = temp % arma::repmat(arma::pow(stdInv, 3),
+      inputSize, 1) * -0.5;
+
+  // Step 3: dl / dxhat * 1 / stdInv + variance * 2 * (x - mu) / m +
+  // dl / dmu * 1 / m.
+  gTemp = (norm.each_slice() % arma::repmat(stdInv,
+      inputSize, 1) +
+      (inputMean.each_slice() % vars * 2)) / (batchSize * higherDimension);
+
+  // Step 4: sum (dl / dxhat * -1 / stdInv) + variance *
+  // (sum -2 * (x - mu)) / m.
+  MatType normTemp = arma::sum(norm.each_slice() %
+      arma::repmat(-stdInv, inputSize, 1) , 2) /
+      (batchSize * higherDimension);
+  gTemp.each_slice() += normTemp;
+}
+
+template<typename MatType>
+void BatchNormType<MatType>::Gradient(
+    const MatType& /* input */,
+    const MatType& error,
+    MatType& gradient)
+{
+  const size_t inputSize = inputDimension;
+
+  gradient.set_size(size + size, 1);
+  arma::Cube<typename MatType::elem_type> errorTemp(
+      const_cast<MatType&>(error).memptr(), inputSize, size,
+      error.n_cols, false, false);
+
+  // Step 5: dl / dy * xhat.
+  MatType temp = arma::sum(arma::sum(normalized % errorTemp, 0), 2);
+  gradient.submat(0, 0, gamma.n_elem - 1, 0) = temp.t();
+
+  // Step 6: dl / dy.
+  temp = arma::sum(arma::sum(errorTemp, 0), 2);
+  gradient.submat(gamma.n_elem, 0, gradient.n_elem - 1, 0) = temp.t();
+}
+
+template<typename MatType>
+void BatchNormType<MatType>::ComputeOutputDimensions()
+{
+  this->outputDimensions = this->inputDimensions;
+  size_t mainAxis = std::min(this->inputDimensions.size() - 1, maxAxis);
+  inputDimension = 1;
+  for (size_t i = 0; i < mainAxis; i++)
+    inputDimension *= this->inputDimensions[i];
+  size = this->inputDimensions[mainAxis];
```

Comment:
So if we set `maxAxis = 2`, won't this mean that we apply batch normalization to the first *three* dimensions?  Correct me if I misunderstood.

Reply:
Just to check, is my statement correct here?  I want to make sure I understand correctly, since this makes a difference for the other suggestions I have left about documentation.